### PR TITLE
[VYMCA-90] Consider user roles in CategoriesController

### PR DESCRIPTION
**Pull request by hamrant**
JIRA - https://fivejars.atlassian.net/browse/VYMCA-90
Basecamp - https://3.basecamp.com/3608022/buckets/17045978/todos/3083961790

## Steps for review:

- [ ] Login as admin
- [ ] Open site in two different browsers or two anonymous windows
- [ ] On both anonymous windows login as dummy user
- [ ] As admin go to `/admin/people`
- [ ] Edit one of the dummy users
- [ ] Uncheck `Virtual YMCA` role and select `Virtual YMCA Trial` and save
- [ ] As result on `/admin/people` you should see

screenshot-carnation openy docksal-2020 10 19-18_55_46

- [ ] As admin go to `/admin/content`
- [ ] Edit `Youth Bootcamp Challenge` video and set there only `Virtual YMCA Trial` role and save

screenshot-carnation openy docksal-2020 10 19-18_59_34

- [ ] As test users check `/virtual-ymca#/categories/video` page
- [ ] User with `Virtual YMCA Trial` role should see only `Bootcamp` category

Screenshot from 2020-10-19 19-06-39

- [ ] User with `Virtual YMCA` role should see only `Yoga` category

Screenshot from 2020-10-19 19-06-59

- [ ] Try to login as another one user and assign for him both roles
- [ ] Check that user can see all categories
- [ ] Test with `Virtual YMCA Premium` should work the same
- [ ] Test this for blogs too `/virtual-ymca#/categories/blog` (edit blogs on content page, to see different categories there)

### CODE REVIEW

File: src/Controller/CategoriesController.php [hunk 1]
```diff
@@ -51,12 +52,37 @@ public static function create(ContainerInterface $container) {
...
     $query->condition('t.vid', 'gc_category');
     $query->condition('tf.status', 1);
+
+    if (!empty($y_roles)) {
+      $or_group = $query->orConditionGroup();
+      foreach ($y_roles as $role) {
+        if ($role == 'virtual_y') {
+          // The rest of roles already contains `virtual_y` string, so we need
+          // more strict condition here.
+          $or_group->condition('nd.field_vy_permission', $role, 'LIKE');
```
Comment (NightHunterSV) [inline src/Controller/CategoriesController.php:76]: Any reason to use 'LIKE' but not '='?

File: src/Controller/CategoriesController.php [hunk 2]
```diff
@@ -51,12 +52,37 @@ public static function create(ContainerInterface $container) {
...
+          $or_group->condition('nd.field_vy_permission', $role . ',%', 'LIKE');
```
Comment (NightHunterSV) [inline src/Controller/CategoriesController.php:77]: Can the `virtual_y` role be not the first one?
If yes -- then we need to check with smth like `'%' . $role . ',%'`

File: src/Controller/CategoriesController.php (hunk 1 above)
Comment (hamrant) [reply src/Controller/CategoriesController.php:76]: DB value can contain multiple roles separated by a comma, for example: `virtual_y,virtual_y_premium`

File: src/Controller/CategoriesController.php (hunk 2 above)
Comment (hamrant) [reply src/Controller/CategoriesController.php:77]: No

File: src/Controller/CategoriesController.php (hunk 1 above)
Comment (NightHunterSV) [reply src/Controller/CategoriesController.php:76]: But for the case above you have no `'%'` in the value which will be compared with the DB value
So tactically it's equal you're using the `'='` operator

Comment (hamrant) [reply src/Controller/CategoriesController.php:76]: in this case yes, we can use `=` instead

Comment (NightHunterSV) [reply src/Controller/CategoriesController.php:76]: Let's do this so -- for the sake of a clarity

Comment (NightHunterSV) [reply src/Controller/CategoriesController.php:76]: Oh, btw, seems it's better to use the same comparison conditions for other roles either:
what if there would be `virtual_y_premium` and `virtual_y_premium_extra` roles?

Comment (hamrant) [reply src/Controller/CategoriesController.php]: Fixed

File: src/Controller/CategoriesController.php (hunk 2 above)
Comment (hamrant) [reply src/Controller/CategoriesController.php]: Fixed

**Review by NightHunterSV — DISMISSED** [review]

File: src/Controller/CategoriesController.php [hunk 3]
```diff
@@ -51,12 +52,31 @@ public static function create(ContainerInterface $container) {
...
   public function list(string $type) {
+    $user = $this->currentUser();
+    $pattern = 'virtual_y';
+    // Get list of vy roles for current user.
+    $y_roles = array_filter($user->getRoles(), function ($role) use ($pattern) {
+      return strpos($role, $pattern) !== FALSE;
+    });
+
     $query = $this->database->select('node__field_gc_video_category', 'n');
+    $query->leftJoin('node_field_data', 'nd', 'n.entity_id = nd.nid');
```
Comment (anpolimus) [inline src/Controller/CategoriesController.php:63]: image
We can do this small fix.

**Review by anpolimus — CHANGES_REQUESTED** [review]

File: src/Controller/CategoriesController.php (hunk 3 above)
Comment (hamrant) [reply src/Controller/CategoriesController.php]: Fixed

**Review by anpolimus — APPROVED** [review]

**Comment by anpolimus** [comment]
Code was tested. Everything is fine.

**Pull request merged**